### ISSUE TITLE
Official Mordenkainen's Tome of Foes Ticket

### ISSUE THREAD

**Issue by dungeonguru** [Sklore/HL_DD_5e_Colab#834]
Anyone working on ToF can use this ticket to update on status, the source for ToF has been added, so please make sure you have downloaded a fresh copy of the source files before you start uploading the new files. 

Please edit this list to indicate if you are working on specific monsters and mark them as complete once done. 

**Tracking Board:** https://github.com/Sklore/HL_DD_5e_Colab/projects/3?add_cards_query=is%3Aopen

* Please use the tracking board to communicate what you are working on and it's stage of completion.

**Comment by LordMorin** [comment]
Tiefling cariants done so far. Working on Demonic Boons atm.

I'm going through the sections one at a time.

**Comment by LordMorin** [comment]
The Tiefling variants are complete. As are the elf subraces, but those were already mostly done in UA. I just had to make the changes made in ToF.
I started on the Duergar.

I was having issues with setting up the Demonic Boons, so I skipped over them for now and will get back to them. I need to figure out how to add a variable stat bonus that can be split between 2 stats . . .

Is there a website than can show me, step by step, how to properly set up a new file for sithub? The last time I tried was a year ago at least and I messed that up. I've been gun shy of working with it since.

**Comment by Sklore** [comment]
There are many ways to upload files to GitHub. Some people connect using apps while I personally literally just use the website. 

I made a doco on how I do it. Might help. 

https://github.com/Sklore/HL_DD_5e_Colab/wiki/How-to-upload-your-changes-to-GitHub

**Comment by LordMorin** [comment]
Thanks. I'll look at it asap. I downloaded the GitrHub Desktop client, but an still trying to make heads or tails of it . . .

**Comment by Sklore** [comment]
FYI I'm working on creating a spreadsheet that will parse the majority of the monsters in very quickly. Will still need to add special abilities manually but this should speed us us significantly.

**Comment by LordMorin** [comment]
<nod> I've been following your posts. :)

I'd working on race related stff then onto boon, etc.

**Comment by Sklore** [comment]
I'm headed to bed. 1am and work starts in a few hours haha. Here's what ive done though in case anyone wants to work on the lookup tables. 

https://drive.google.com/file/d/1V3xSgOAQo0TGmb220Ee8Ro1q8zLVnGb2/view?usp=sharing

**Comment by LordMorin** [comment]
kk, have a good one. Only noon here. :)

**Comment by Sklore** [comment]
Parser is working for the foundation of the monsters :)
![image](https://user-images.githubusercontent.com/8046697/40877195-cf1affbe-66bb-11e8-8cc2-0afdb68b3acb.png)

**Comment by Sklore** [comment]
Ok the foundation for all monsters is now complete. 

Things that should be complete include:
Name, Unique Name, Size, Type, Alignment, AC, Hit Points, Speeds, Stats, Saving Throws (check results), Skills (check results), Condition Immunities, Senses, Challenge. 

Everything else will need to be completed manually. 

Hopefully this helps speed the process up. 

![image](https://user-images.githubusercontent.com/8046697/40877686-105ed308-66c4-11e8-9485-1a01107013f1.png)

**Comment by LordMorin** [comment]
Sorry about the delays. RL issues have delayed my work on MToF's subraces and stuff. I've restarted my work today and hope to have the subraces done soon.

**Comment by LordMorin** [comment]
Update
- Added Tiefling variants
- Added Duegar
- Added Gith (Githyanki & Githzerai)

**Comment by LordMorin** [comment]
Update

Added Depp Gnome & Svirfneblin Magic feat
- these are same as the Deep Gnome in Elemental Evil as far as I can te;;

**Comment by LordMorin** [comment]
Update

- Added Deep Gnome
- Added new & updated deities

**Comment by LordMorin** [comment]
Is it possible for someone else to download and test my branch before I attempt to merge it?

**Comment by LordMorin** [comment]
Update
- missed adding new & updated Drow deities - fixed

**Comment by LordMorin** [comment]
Update
- missed adding new & updated Drow deities - fixed

**Comment by LordMorin** [comment]
Well, I have my own branch with the file 'COM_5ePack_MTOF - Races' in it. I have been testing it between my main system, laptop, and iPad and I think I have resolved all ther issues I can think of.

So how, or should I, merge it with the main branch? Right now the file is only in my Mergon branch.

**Comment by dungeonguru** [comment]
Before you merge, look at ra5CToFNeR (Necrotic Resistance) - you introduced a new source 'EEvil" from what I think is your personal files - change it to the new MTOF source and your files look otherwise clean enough to merge in as you have only a single large file that has changes.   In order to merge you need to have your branch selected on the <> Code tab. 
You click on the New Pull Request button beside where you code is selected. 
On the next screen it will show whether or not github thinks it is safe to merge. Put this ticket number in the comments section somewhere and it will autolink the merge to this ticket (like type ticket #834 - MToF).
Under that comment box will be a green Create Pull Request button. Click it. 
At this point you can wait till someone grabs the pull request and then merges it for your or if you're good with your code you can click on a button on this next page that says Merge Pull Request and follow that dialog (which will also allow you to delete your branch if need be).

**Comment by LordMorin** [comment]
Ok. I did the pull. I then to my Code branch and updated my fixed code.

So do I now just do the Merge Pull Request?

**Comment by LordMorin** [comment]
I did the Create Merge. let me know if I messed it up. :)

**Comment by dungeonguru** [comment]
I think you did good.  A tip/trick for you - when you go to the code page after you merge, the top line should show the merge like this:

> LordMorin Merge pull request #849 from Sklore/mergon  …
> Latest commit f341317  3 minutes ago

If you click on the commit hash number (f341317) it will take you to a page that shows exactly what changed. Your change looks like it uploaded a single file and when I do a Ctrl-F and search for that EEvil source it doesn't show up on the page.

**Comment by LordMorin** [comment]
I'd already created the Necrotic Resistance ability for a custom race of my own. I simply copied that5 for MToF but forgot to change the source.

It's always little things like that that bite you on the a**.

**Comment by MikevanSchijndel** [comment]
Hi, 

I followed the instructions on RW forum. I am still confused on how to see what monster I can work on. Only one of them has (complete) after its name. So, is that the only monsters thats completed? Can I start at the top of the list? 

Thanks, 

Mike

**Comment by MikevanSchijndel** [comment]
Working on Adult Kruthik

**Comment by Sklore** [comment]
I've upped your permissions Mike. This means you should be able to edit the original post at the top to indicate what you are working on and post completion status. One's with the COMPLETE status are the ones that have been finished. Everything else is fair game. 

I recommend identifying a single chunk to work on. All of the A's for example. Obviously dependant on available time.

**Comment by MikevanSchijndel** [comment]
Thanks I was just about to ask for permission! 

I got plenty of time. It's easy work and I love having the monsters in HL so I think I will throw myself at the list in the coming days.

**Comment by Sklore** [comment]
Bugger, that must be special access for me since I own the GitHub site. 

That's everything we need to do in the editor. There is more to do in Hero Lab itself (add weapons and armour) but we will get a single person to do that once everything else is ready.

**Comment by Sklore** [comment]
Since you guys can't edit the primary list I have created a tracking sheet in Google Docs. Please use this to track who is working on what and it's completion status:

https://docs.google.com/spreadsheets/d/1yOAWffNl19J6Tp8-2MkMw0-uQoSQUGPqobWi-AaFHRo/edit?usp=sharing

**Comment by dungeonguru** [comment]
I started in on the monsters starting with G. For the ones that are duplicates I added the tag Helper.Obsolete and changed the monster name to ### (duplicate). I did this so future merges might not re-add incomplete characters - or that's my hope. I also am updating the MM file for these duplicates to include the new MTOF source.

**Comment by LordMorin** [comment]
Ok, I am working on Demonic Boons for MToF. Each on grants stats similar to this: "You gain up to a +4 bonus to Strength, Wisdom, or both.".

What I am looking at is the best way to implement it. You can add a +1 to +4 split between 1 to 2 stats. Any suggestions on the best way to add this ability?

**Comment by dungeonguru** [comment]
The only way I can think to variable ability stats up would be to copy the Ability Score adjustments. I think there is a way to set minimum and maximum amounts there. I've been trying to figure out the best way to implement Demonic Boons in my own files and can't figure out a better way .

**Comment by LordMorin** [comment]
<nod>

Once I get that part worked out, the rest is wasy. :(

I've tried using Bootstrapped Abilities, racial abilities so far. My next path is 2 try adding 2 abilities, one for each stat. I am not sure if you can bootstrap adjustments.

I'm hoping ShadowChemosh sees this and has an idea. . .

**Comment by shadowchemosh** [comment]
> What I am looking at is the best way to implement it. You can add a +1 to +4 split between 1 to 2 stats. Any suggestions on the best way to add this ability?

I would create a pair of helper Ability Things (one for Wis other for Str). Then use the Activation amount fields of 0 to 4 to control the amount going to each ability score.  Bootstrap these to the your Ability/Magic item.  Have the main Pick with the bootstrapped Abilities do the validation of only having a combined amount of 4 points.  Easy way would be to have the abilities access the root transition and update say abValue with the total amount used.

**Comment by Fenris447** [comment]
I’m throwing my hat in and helping. Got my toes wet with the Phoenix (only P monster) last night.

Is MikevanSchijndel still around? It looks like he completed a bunch of monsters according to the tracker, but I don’t see his work uploaded anywhere here.

**Comment by MikevanSchijndel** [comment]
Hi Fenris, 

Yes, I am still around. However I am in the middle of renovating my  house into one huge gaming room. :D I will try and upload whatever I have when I have a bit more time. I'll need some help to determine which files I need to upload though. It's been a while.

**Comment by anthonydeaver** [comment]
I'd be happy to help out where I can. I can start with the F's if no one else is working on it.

**Comment by Fenris447** [comment]
L, P, and Z are programmed and I’ve marked them as such in the Project.

Here’s the bigger thing I’ve done: Between every book we haven’t done all the monsters for (HotDQ, RoT, PotA, MToF, W:DH, W:DotMM, GGtR, GoS), I pulled a list of all monster abilities.

Any non-unique abilities, where the only difference is a name or a number, I have programmed in. You only need to select this version of the ability then assign the new values for abText, abValue, etc.

I’m going to upload all these files, create a tutorial page in this wiki, and present the full list so you can reference them as you program monsters.

Here’s the problem: I’ll be uploading a new MTOF with these abilities. If you download that, it’ll overwrite any work you’ve done already. To prevent that, we need to merge the files. I can do this manually if people put their work into this thread in a .zip file. I’m not sure there’s an easier way than that. But once we get that done and we all start work off a new, merged file with those modular abilities, this should go much faster.

What does everyone think?

**Comment by Fenris447** [comment]
I have a problem uploading my work for MTOF. It looks like the Master has the basic monster info, but the MTOF branch doesn’t yet. I added my new MTOF file to the branch, but am told it’s unable to merge with the Master. GitHub won’t open the interface to review the issue, and I don’t think I’d know what to do with it if it did.

I manually compared the Master’s code with my own, and I don’t see any issues with it.

Should I replace the MTOF branch’s code with what’s in the Master, merge it, then add my code back to the branch and merge again? Seems like the best solution, but I can hold off if someone tells me to.

**Comment by LordMorin** [comment]
I think the only work I've done on MToF was the new subraces? I know I was trying to do Demonic Boons then I had my stroke. I don't think I ever went back to them. I completely forgot till now. :(

I know I have done no work on the monsters though.

**Comment by Fenris447** [comment]
I was able to get the conflict resolution tool working. By using Edge. Never thought that would be the solution to anything. But now my stuff is merged in and good to go!

**Comment by LordMorin** [comment]
Who would have thought a Microsoft product would solve a problem instead of causing one. :)

**Comment by dungeonguru** [comment]
Well, since github was bought by microsloth last year.... expect any decent browser functions to be really quirky is all I'll say.

**Comment by Fenris447** [comment]
How’s it coming for everyone else? I’ve re-written the ability text for all monsters S, T, and W, and am in the process of adding them in through the editor. I have a running stock file as I go. Is anyone held up on a problem we could solve together?

Also, if anyone has monsters ready to integrate into the files but doesn’t want to go through the github process, let me know. I’m happy to just take your file and handle that side of things.

**Comment by MikevanSchijndel** [comment]
Hi Fenris!

I haven't done any work but I still have the work I did last year.  I can
send it to you if you tell me what I'm looking for and send me an email
adress to send it to. :)

Thanks,

Mike

Op ma 8 jul. 2019 om 19:28 schreef Fenris447 <notifications@github.com>:

**Comment by Fenris447** [comment]
Mike, if you can send your local copy of COM_5ePack_MTOF - Races.user to Fenris447@gmail.com, I can take what you’ve done and integrate it. Thanks!

**Comment by Fenris447** [comment]
FYI I’m about 50% done creating the special abilities of all the remaining monsters. I’ve been working on a way to dump them in, similar to how we did the monsters to begin with. If anyone has work that isn’t already uploaded here, please email your copy of COM_5ePack_MTOF - Monsters.user to me at Fenris447@gmail.com. I want to get this out before the next big book release.

**Comment by AmbushXXVI** [comment]
So, I've finally gotten some things straightened out in my personal life and I'd like to get back to trying to get this out the door. Is the current state accurate as far as if I pick a letter that doesn't seem done, I'm not going to be duplicating anyone's work?

**Comment by Fenris447** [comment]
I’ve been tackling a bit of everything. The stuff I’ve done the least on is N and O, followed by M. If you can knock those out, it would really help!

**Comment by AmbushXXVI** [comment]
Gotchu. I'll get started on N.

**Comment by AmbushXXVI** [comment]
Holy crap, spellcasting is a pain in the ass! It only took me about 90m to get the Nagpa done. 

But it's done.

**Comment by Bombedudu** [comment]
I am almost done with the elemental myrmidons. How do I add their weapons?

**Comment by Fenris447** [comment]
You don’t add them in the editor. They’ll go in when we make a stock file. Stock files are basically portfolios using each monster as the race for a hero, adding in the appropriate items and spells, and then saving it as a .stock file. That’s what allows Hero Lab to pull these monsters into the Encounter Builder.

**Comment by Fenris447** [comment]
I’m making progress on MTOF daily at this point. I have programmed in all but about 40 monster abilities and they’re in progress. I’m also making the stock versions as I go, assigning those abilities, adding attacks and spellcasting, etc. one by one. 

I’m not putting a date on it. But the last community pack release was 2/2/18 and I think two years between releases is too long.

**Comment by dungeonguru** [comment]
I've advocated in the past to push at least quarterly typo and other minor updates even if there are pending items. 

If that's something that you want to take over make sure you read the #5 of the wiki, update the #3 page with the pending changes (review the commit log from the master merges to see if any undocumented changes are snuck in)  and get in contact with ShadowChemosh to coordinate him making the release from the main branch. Push what you have merged to main and then in 3 months push again. 

I would also suggest testing the main branch on a fresh install of HLC BEFORE having ShadowChemosh commit though. Sometimes local or personal changes sneak into the repository and create a major headache.

**Comment by Fenris447** [comment]
Yeah I’ll definitely be testing the whole thing on a fresh install before I push the release.

I’m getting to a point where I have a steady workflow for monsters. While I’d love to commit to quarterly updates, I just don’t know if I’ll be able to keep up. I don’t know anyone else at this point who is actively working on the Community Pack, and dwindling support from LWD is only going to make it worse.

That being said, I’ll keep going so long as the time spent:value ratio keeps HL the best option.

Also I’m at about 80% on these monsters. Barring anything crazy, I should have it done before February. The only thing that won’t be done is Demonic Boons, which are pretty low-priority for me.

**Comment by Fenris447** [comment]
I can happily announce that all MTOF monsters have been programmed and had stock files created.

The only things unfinished are some lair actions, the clockwork adjustments, and Demonic boons. I’m not going to hold back the release just for those.

Next steps will be for me to upload all my changes (it’s a lot) to my brunch, catalogue them on the pending release notes, and do a test with a clean install.

**Comment by Fenris447** [comment]
Now that the core Monsters and Races are in, we only have a few more items to do to make MTOF feature-complete:

- Demonic Boons - I believe @LordMorin made some progress on these, but they’re pretty complicated 
- Clockwork Adjustments - @dungeonguru did you do any more than what’s already in the files? You’ve got the framework but most of the individual selections aren’t programmed
- Lair Actions - I’ll admit it; I wanted to get the release out sooner rather than later, so I put these off. I’ll get them done for our next release

**Comment by LordMorin** [comment]
I'd completely forgotten about demonic boons. I got into Eberron stuff about the same time and never thought back.
I never did come up with a way to implement the Demonic Boons that would work.

**Comment by dungeonguru** [comment]
I'm not sure what you mean on the clockwork. In order to add a clockwork enhancement or malfunction you need to do it manually. You load up a critter and go to the adjustments tab, select the clockwork enhancement adjustment found in the Race section and then in the dropdown choose which one you want.  It should bootstrap the racial ability to the critter and then run the scripts on those abilities. Check out the p5CCloMal and p5CCloEnh things and you should see where they call out a script to load the racial abilities based on the dropdown index. 

I don't run the community files anymore but I checked and my personal files have the same logic and my adjustments are working fine.

**Comment by Fenris447** [comment]
@dungeonguru  Sorry, I was vague. The enhancements/malfunctions are in and from what I could tell, you had started programming them to have their effects once chosen. IE if you chose “Camouflaged” from the drop down, it bootstraps the Camouflaged ability, which grants the creature proficiency in stealth. But it looks like, in the files here, only the first three of those to-be-bootstrapped Enhancement abilities were created.

My question was whether or not you had done more than the overall adjustments and those first three abilities, and just didn’t get around to uploading the rest. No worries if not; I know you’re out of the game. I just don’t want to redo work if it already exists on your hard drive.

**Comment by dungeonguru** [comment]
Well, I got to looking now that I'm able to search without getting interrupted.  I think you're referring to the ability: ra5CIroCobEn "Clockwork Enhancement" in the community file - that's an orphan and can be hidden or deleted with the next release, nothing bootstraps or calls it or the other orphan ra5CIroCobMa . I didn't finish those because the adjustment logic wouldn't work with nesting the abilities like that and then I complicated it by forgetting to delete them. You can add a Helper.Obsolete tag to them if you want. 

If you look at the two adjustments (p5CCloEnh, p5CCloMal)  you can see that they actually do bootstrap all 10 enhancements and malfunctions respectively and when I do a search I can see all of those racials added in. Some of these don't have scripting because the ability can't be replicated.  I also see where I didn't go back in and give camoflauged and sensors a better ID (I was doing Iron Cobra initially). 

Camoflauged - ra5CIroCam
Sensors - ra5CIroSen
Improved AC - ra5CCloArm
Increased Speed - ra5CCloSpd
Reinforced Construction - ra5CCloRein
Self Repairing - ra5CCloRep   (no script, HL doesn't have healing or per round effects)
Sturdy Frame - ra5CCloFra (hit point scripts don't seem to work right 100% of time so I didn't finish this)
Suction - ra5CCloSuc  !This one is confusing since the script doesn't work and the adjustment looks like we're adding xClimb with a value of 30, however the ra5CCloSuc is needed since it has the abValue
Vocal Resonator - ra5CCloVoc
Water Propulsion - ra5CCloSwm  !same issues as suction

Faulty Sensors - ra5CCloFaSe (no script, HL doesn't have automatic per round funcitons like a VTT)
Flawed Targeting - ra5CCloFlaTar (no script/VTT)
Ground Fault - ra5CCloGroFau (scripted)
Imprinting Loop - ra5CCloImp (no script, HL doesn't automatically roll dice per round for anything)
Leaking Lubricant - ra5CCloLub (no script, HL doesn't auto assign exhaustion levels per round)
Limited Steering - ra5CCloSte (no script, HL doesn't actually track movement per round)
Overactive Sense of Self-Preservation - ra5CCloOSS (no script, HL isn't a VTT)
Overheats - ra5CCloHeat (no script, not a VTT)
Rusty Gears - ra5CCloRust (no script, not a function of HL)
Weak Armor - ra5CCloWeak (scripted)

**Comment by Fenris447** [comment]
You’re absolutely right. I totally missed them.  Sorry for the hassle, and thanks for the great work!

**Comment by Fenris447** [comment]
In other news, I’m working on the boons. I’m looking at doing the stat adjustments as arrays. It’s not terribly pretty, but it lets us get at all the options without super-complicated programming.

**Comment by LordMorin** [comment]
I'd been looking at a pair of bootstrapped abilities just to cover the stats. The problem was multiple stats adding up to +1 to +3 (I think, it might have been +5).

Arrays were never my strong suit.

**Comment by Fenris447** [comment]
I’m doing it like this. Arrays with all options (just two here for demonstration, there’s more for each one for every combination available under a particular demon):

<arrayval field="usrArray" index="0" value="+1 STR"/>
<arrayval field="usrArray" index="1" value="+1 STR, +1 WIS"/>


Then there are scripts on each boon that apply the chosen stat boosts based on the array option chosen:

if (field[usrIndex].value = 0) then
#skillbonus[aSTR] += 1
elseif (field[usrIndex].value = 1) then
#skillbonus[aSTR] += 1
#skillbonus[aWIS] += 1

I haven’t actually tested the code yet, but I don’t see why it wouldn’t function.

**Comment by LordMorin** [comment]
Good luck and I hope everything works. I tried many different things with no luck.

**Comment by Fenris447** [comment]
The remaining lair actions are completed and programmed in. I’ve added the clockwork adjustments to the appropriate monsters in the .stocks.

All that’s left is the Boons and we can finally close the book on MTOF!

**Comment by Fenris447** [comment]
Finally, 2.5 years after release, we can close this book. I’ve completed and uploaded the Demonic Boons (renamed to Fiendish Boons and Cults). This includes all the fiend cults, demonic boons, other cults, and variant cambion features. There’s an adjustment that enables a configurable to pick from any and all of these. Picking the cult or boon in the primary field opens up the secondary with all the features from that boon/cult. You can select multiple boons and cults if you wish, and the cambion features can also be chosen there.
